{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "5a0b1613",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Daniel Castaneda Mogollon, PhD\n",
    "#January 14th, 2025\n",
    "#This script was made to fetch the 16S sequences from Sanger and WGS and merge them with the most updated\n",
    "#version of GTDB from the Master file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "dd5de1a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#importing packages\n",
    "import pandas as pd\n",
    "import os\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "e41a470f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Setting path\n",
    "path = \"/Users/danielcm/Desktop/Sycuro/Projects/Diabetes/\"\n",
    "os.chdir(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "2be1009b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Defining files\n",
    "master_file = pd.read_csv(\"WGS_all_runs/FINAL_report_WGS021224.csv\")\n",
    "fasta_file = \"16S_data/16S_from_both/all_approaches_merged_rmdup.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "6db4f472",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This sections corresponds to the first DADA2 file that needs the whole taxonomy as part of its header\n",
    "master_file_filtered = master_file.loc[master_file['Selected for Downstream']=='Yes',\n",
    "                                       ['Sample number','Classification']] #filters selected for downstream and prints these two columns only into a new df\n",
    "master_file_filtered['Sample number'] = master_file_filtered['Sample number'].apply(lambda x: f'{int(x):03d}') #This ensures that I have always three digits\n",
    "sample_classification_list = master_file_filtered['Classification'].tolist() #transforms column to list\n",
    "sample_number_list = master_file_filtered['Sample number'].tolist() #transforms colum to list\n",
    "updated_records = [] #creates an empty list of updated records to be added (in this case updated header)\n",
    "with open(fasta_file,'r') as fasta:\n",
    "    for record in SeqIO.parse(fasta,\"fasta\"):\n",
    "        header = record.id\n",
    "        sequence = str(record.seq)\n",
    "        header_sample_number = header.split(\"_\")[0] #makes sure I only get the first part '>seqxxx'\n",
    "        header_sample_number = header_sample_number.split(\"seq\")[1] #makes sure I only get the number 'xxx'\n",
    "        if header_sample_number in sample_number_list: #if the header from the original fasta file is in the excel file then do the following\n",
    "            index = sample_number_list.index(header_sample_number) #saves the index number\n",
    "            new_header = f\"{sample_classification_list[index]}\" #gets the classification corresponding to the index number\n",
    "            record.id = new_header #updates new header to the existing record\n",
    "        updated_records.append(record)\n",
    "\n",
    "with open(\"dada2_new_updatad_taxonomy.fasta\",\"w\") as updated_fasta:\n",
    "    SeqIO.write(updated_records,updated_fasta,\"fasta\") #writes new updated fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "2ed84d8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seq001_1 Akkermansia muciniphila\n",
      "seq002_1 Alistipes finegoldii\n",
      "seq003_1 Alistipes onderdonkii\n",
      "seq004_1 Bacteroides stercoris\n",
      "seq005_1 Bacteroides thetaiotaomicron\n",
      "seq006_1 Bacteroides uniformis\n",
      "seq007_2 Phocaeicola vulgatus\n",
      "seq007_3 Phocaeicola vulgatus\n",
      "seq008_1 Bifidobacterium longum\n",
      "seq009_5 Bifidobacterium longum\n",
      "seq010_1 Bifidobacterium pseudocatenulatum\n",
      "seq011_2 Enterocloster bolteae\n",
      "seq012_5 Enterocloster citroniae\n",
      "seq013_2 Sarcina perfringens\n",
      "seq014_1 Eggerthella lenta\n",
      "seq015_2 Eisenbergiella tayi\n",
      "seq016_1 Enterococcus_B durans\n",
      "seq016_2 Enterococcus_B durans\n",
      "seq016_4 Enterococcus_B durans\n",
      "seq016_6 Enterococcus_B durans\n",
      "seq016_7 Enterococcus_B durans\n",
      "seq017_4 Escherichia coli\n",
      "seq018_1 Escherichia coli\n",
      "seq021_1 Hungatella effluvii\n",
      "seq022_1 Parabacteroides distasonis\n",
      "seq023_2 Parabacteroides merdae\n",
      "seq025_1 Staphylococcus hominis\n",
      "seq025_3 Staphylococcus hominis\n",
      "seq026_2 Streptococcus sp001556435\n",
      "seq027_1 Sutterella wadsworthensis\n",
      "seq027_2 Sutterella wadsworthensis\n",
      "seq029_1 Akkermansia massiliensis\n",
      "seq029_3 Akkermansia massiliensis\n",
      "seq029_4 Akkermansia massiliensis\n",
      "seq030_1 Alistipes finegoldii\n",
      "seq032_1 Bacteroides stercoris\n",
      "seq034_2 Bacteroides uniformis\n",
      "seq036_1 Bacteroides uniformis\n",
      "seq036_2 Bacteroides uniformis\n",
      "seq037_2 Phocaeicola vulgatus\n",
      "seq038_2 Bacteroides xylanisolvens\n",
      "seq039_1 Bifidobacterium adolescentis\n",
      "seq040_1 Bifidobacterium longum\n",
      "seq041_1 Enterocloster bolteae\n",
      "seq043_3 Sarcina perfringens\n",
      "seq044_2 Eggerthella lenta\n",
      "seq045_1 Eggerthella lenta\n",
      "seq046_1 Eisenbergiella massiliensis\n",
      "seq048_2 Eisenbergiella tayi\n",
      "seq049_1 Enterococcus faecalis\n",
      "seq050_1 Enterococcus faecalis\n",
      "seq051_1 Escherichia coli\n",
      "seq051_2 Escherichia coli\n",
      "seq053_1 Hungatella effluvii\n",
      "seq054_1 Hungatella sp005845265\n",
      "seq055_1 Parabacteroides distasonis\n",
      "seq056_1 Parabacteroides merdae\n",
      "seq056_3 Parabacteroides merdae\n",
      "seq057_1 Pseudoflavonifractor_A sp022772585\n",
      "seq058_1 Staphylococcus hominis\n",
      "seq058_2 Staphylococcus hominis\n",
      "seq059_1 Streptococcus sp001556435\n",
      "seq060_1 Sutterella wadsworthensis\n",
      "seq061_1 Akkermansia muciniphila\n",
      "seq063_2 Bacillus subtilis\n",
      "seq064_3 Phocaeicola dorei\n",
      "seq065_2 Phocaeicola vulgatus\n",
      "seq066_1 Bifidobacterium dentium\n",
      "seq068_1 Enterocloster bolteae\n",
      "seq069_1 Alitiscatomonas sp900066535\n",
      "seq070_1 Enterocloster clostridioformis\n",
      "seq071_1 Otoolea symbiosa\n",
      "seq073_1 Eggerthella lenta\n",
      "seq074_1 Eisenbergiella tayi\n",
      "seq074_2 Eisenbergiella tayi\n",
      "seq075_1 Eisenbergiella tayi\n",
      "seq076_1 Enterococcus_A avium\n",
      "seq077_1 Enterococcus_A avium\n",
      "seq078_1 Escherichia coli\n",
      "seq079_1 Faecalibacterium longum\n",
      "seq080_1 Flavonifractor plautii\n",
      "seq081_1 Holdemania filiformis\n",
      "seq086_1 Hungatella effluvii\n",
      "seq083_1 Parabacteroides distasonis\n",
      "seq083_2 Parabacteroides distasonis\n",
      "seq084_1 Hungatella effluvii\n",
      "seq085_1 Ruminococcus_B gnavus\n",
      "seq087_1 Staphylococcus hominis\n",
      "seq088_1 Collinsella sp902472315\n",
      "seq089_2 Akkermansia muciniphila\n",
      "seq090_3 Alistipes finegoldii\n",
      "seq091_1 Alistipes finegoldii\n",
      "seq092_1 Cytobacillus undefined\n",
      "seq093_2 Phocaeicola vulgatus\n",
      "seq094_1 Bifidobacterium longum\n",
      "seq095_1 Bifidobacterium pseudocatenulatum\n",
      "seq096_1 Enterocloster bolteae\n",
      "seq097_1 Enterocloster citroniae\n",
      "seq098_1 Enterocloster clostridioformis\n",
      "seq099_1 Enterocloster alcoholdehydrogenati\n",
      "seq100_2 Eggerthella lenta\n",
      "seq101_1 Eisenbergiella porci\n",
      "seq102_1 Eisenbergiella tayi\n",
      "seq104_1 Escherichia coli\n",
      "seq105_2 Faecalibacterium longum\n",
      "seq107_1 Holdemania filiformis\n",
      "seq108_1 Hungatella effluvii\n",
      "seq110_1 Parabacteroides distasonis\n",
      "seq111_1 Parabacteroides distasonis\n",
      "seq112_1 Agathobacter faecis\n",
      "seq113_1 Ruminococcus_B gnavus\n",
      "seq115_1 Instestinimonas_A celer\n",
      "seq118_1 Roseburia_C amylophila\n",
      "seq119_1 Alistipes finegoldii\n",
      "seq120_1 Staphylococcus epidermidis\n",
      "seq100 Eggerthella lenta\n",
      "seq101 Eisenbergiella porci\n",
      "seq102 Eisenbergiella tayi\n",
      "seq103 Enterococcus_A avium\n",
      "seq104 Escherichia coli\n",
      "seq105 Faecalibacterium longum\n",
      "seq106 Flavonifractor plautii\n",
      "seq107 Holdemania filiformis\n",
      "seq108 Hungatella effluvii\n",
      "seq109 CHH4-2 sp018378255\n",
      "seq010 Bifidobacterium pseudocatenulatum\n",
      "seq110 Parabacteroides distasonis\n",
      "seq111 Parabacteroides distasonis\n",
      "seq112 Agathobacter faecis\n",
      "seq113 Ruminococcus_B gnavus\n",
      "seq115 Instestinimonas_A celer\n",
      "seq116 Bifidobacterium longum\n",
      "seq117 Bacteroides uniformis\n",
      "seq118 Roseburia_C amylophila\n",
      "seq119 Alistipes finegoldii\n",
      "seq011 Enterocloster bolteae\n",
      "seq120 Staphylococcus epidermidis\n",
      "seq012 Enterocloster citroniae\n",
      "seq013 Sarcina perfringens\n",
      "seq014 Eggerthella lenta\n",
      "seq015 Eisenbergiella tayi\n",
      "seq016 Enterococcus_B durans\n",
      "seq017 Escherichia coli\n",
      "seq018 Escherichia coli\n",
      "seq019 Flavonifractor plautii\n",
      "seq001 Akkermansia muciniphila\n",
      "seq020 Flavonifractor plautii\n",
      "seq021 Hungatella effluvii\n",
      "seq022 Parabacteroides distasonis\n",
      "seq023 Parabacteroides merdae\n",
      "seq025 Staphylococcus hominis\n",
      "seq026 Streptococcus sp001556435\n",
      "seq027 Sutterella wadsworthensis\n",
      "seq028 Collinsella sp902362275\n",
      "seq029 Akkermansia massiliensis\n",
      "seq002 Alistipes finegoldii\n",
      "seq030 Alistipes finegoldii\n",
      "seq031 Alistipes onderdonkii\n",
      "seq032 Bacteroides stercoris\n",
      "seq033 Bacteroides stercoris\n",
      "seq034 Bacteroides uniformis\n",
      "seq036 Bacteroides uniformis\n",
      "seq037 Phocaeicola vulgatus\n",
      "seq039 Bifidobacterium adolescentis\n",
      "seq003 Alistipes onderdonkii\n",
      "seq040 Bifidobacterium longum\n",
      "seq041 Enterocloster bolteae\n",
      "seq042 Enterocloster sp005845215\n",
      "seq043 Sarcina perfringens\n",
      "seq044 Eggerthella lenta\n",
      "seq045 Eggerthella lenta\n",
      "seq046 Eisenbergiella massiliensis\n",
      "seq047 Eisenbergiella massiliensis\n",
      "seq048 Eisenbergiella tayi\n",
      "seq049 Enterococcus faecalis\n",
      "seq004 Bacteroides stercoris\n",
      "seq050 Enterococcus faecalis\n",
      "seq051 Escherichia coli\n",
      "seq052 Flavonifractor sp000508885\n",
      "seq053 Hungatella effluvii\n",
      "seq054 Hungatella sp005845265\n",
      "seq055 Parabacteroides distasonis\n",
      "seq056 Parabacteroides merdae\n",
      "seq057 Pseudoflavonifractor_A sp022772585\n",
      "seq058 Staphylococcus hominis\n",
      "seq059 Streptococcus sp001556435\n",
      "seq005 Bacteroides thetaiotaomicron\n",
      "seq060 Sutterella wadsworthensis\n",
      "seq061 Akkermansia muciniphila\n",
      "seq063 Bacillus subtilis\n",
      "seq064 Phocaeicola dorei\n",
      "seq065 Phocaeicola vulgatus\n",
      "seq066 Bifidobacterium dentium\n",
      "seq067 Bifidobacterium longum\n",
      "seq069 Alitiscatomonas sp900066535\n",
      "seq006 Bacteroides uniformis\n",
      "seq070 Enterocloster clostridioformis\n",
      "seq071 Otoolea symbiosa\n",
      "seq072 Eggerthella lenta\n",
      "seq073 Eggerthella lenta\n",
      "seq074 Eisenbergiella tayi\n",
      "seq075 Eisenbergiella tayi\n",
      "seq076 Enterococcus_A avium\n",
      "seq077 Enterococcus_A avium\n",
      "seq078 Escherichia coli\n",
      "seq079 Faecalibacterium longum\n",
      "seq007 Phocaeicola vulgatus\n",
      "seq080 Flavonifractor plautii\n",
      "seq081 Holdemania filiformis\n",
      "seq082 Hungatella effluvii\n",
      "seq083 Parabacteroides distasonis\n",
      "seq084 Hungatella effluvii\n",
      "seq085 Ruminococcus_B gnavus\n",
      "seq086 Hungatella effluvii\n",
      "seq087 Staphylococcus hominis\n",
      "seq088 Collinsella sp902472315\n",
      "seq089 Akkermansia muciniphila\n",
      "seq008 Bifidobacterium longum\n",
      "seq090 Alistipes finegoldii\n",
      "seq093 Phocaeicola vulgatus\n",
      "seq094 Bifidobacterium longum\n",
      "seq095 Bifidobacterium pseudocatenulatum\n",
      "seq096 Enterocloster bolteae\n",
      "seq097 Enterocloster citroniae\n",
      "seq098 Enterocloster clostridioformis\n",
      "seq099 Enterocloster alcoholdehydrogenati\n",
      "seq009 Bifidobacterium longum\n",
      "seq038 Bacteroides xylanisolvens\n",
      "seq068 Enterocloster bolteae\n",
      "seq091 Alistipes finegoldii\n",
      "seq092 Cytobacillus undefined\n",
      "ID: seq001_1 Akkermansia muciniphila\n",
      "Name: seq001_1\n",
      "Description: seq001_1\n",
      "Number of features: 0\n",
      "Seq('ATGGAGAGTTTGATTCTGGCTCAGAACGAACGCTGGCGGCGTGGATAAGACATG...TTT')\n"
     ]
    }
   ],
   "source": [
    "#This section corresponds to the second DADA2 file that needs just the ID Genus Species for the 'AssignSpecies'\n",
    "#code\n",
    "master_file_filtered2 = master_file.loc[master_file['Selected for Downstream']=='Yes',\n",
    "                                        ['Sample number','GTDBtk Species Classification (pplacer)']]\n",
    "master_file_filtered2['Sample number'] = master_file_filtered2['Sample number'].apply(lambda x: f'{int(x):03d}') #This ensures that I have always three digits\n",
    "sample_classification_list2 = master_file_filtered2['GTDBtk Species Classification (pplacer)'].tolist()\n",
    "sample_number_list2 = master_file_filtered2['Sample number'].tolist()\n",
    "updated_records2 = []\n",
    "with open(fasta_file,'r') as fasta:\n",
    "    for record2 in SeqIO.parse(fasta, \"fasta\"):\n",
    "        header2 = record2.id\n",
    "        sequence2 = str(record2.seq)\n",
    "        header_sample_number2 = header2.split(\"_\")[0]\n",
    "        header_sample_number2 = header_sample_number2.split(\"seq\")[1]\n",
    "        if header_sample_number2 in sample_number_list2:\n",
    "            index2 = sample_number_list2.index(header_sample_number2)\n",
    "            new_header2 = f\"{header2} {sample_classification_list2[index2]}\"\n",
    "            record2.id = new_header2\n",
    "            print(header2 +\" \"+ sample_classification_list2[index2])\n",
    "        updated_records2.append(record2)\n",
    "print(updated_records2[0])\n",
    "with open(\"dada2_assign_species.fasta\",\"w\") as updated_fasta2:\n",
    "    SeqIO.write(updated_records2, updated_fasta2, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc08c0fe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
